src/pip/_internal/commands/cache.py

```diff
@@ -204,22 +206,55 @@ def _find_http_files(self, options: Values) -> List[str]:
     def _find_wheels(self, options: Values, pattern: str) -> List[str]:
         wheel_dir = self._cache_dir(options, "wheels")
 
-        # The wheel filename format, as specified in PEP 427, is:
+        # The wheel filename format, as originally specified in PEP 427, is:
         #     {distribution}-{version}(-{build})?-{python}-{abi}-{platform}.whl
         #
         # Additionally, non-alphanumeric values in the distribution are
         # normalized to underscores (_), meaning hyphens can never occur
         # before `-{version}`.
         #
         # Given that information:
-        # - If the pattern we're given contains a hyphen (-), the user is
-        #   providing at least the version. Thus, we can just append `*.whl`
-        #   to match the rest of it.
-        # - If the pattern we're given doesn't contain a hyphen (-), the
-        #   user is only providing the name. Thus, we append `-*.whl` to
-        #   match the hyphen before the version, followed by anything else.
+        # - If the pattern we're given is not a glob:
+        #   + We attempt project name normalization such that pyyaml matches PyYAML,
+        #     etc. as outlined here:
+        #     https://packaging.python.org/specifications/name-normalization.
+        #     The one difference is we normalize non-alphanumeric to `_` instead of `-`
+        #     since that is how the wheel filename components are normalized
+        #     (facepalm).
+        #   + If the pattern we're given contains a hyphen (-), the user is providing
+        #     at least the version. Thus, we can just append `*.whl` to match the rest
+        #     of it if it doesn't already glob to the end of the wheel file name.
+        #   + If the pattern we're given is not a glob and doesn't contain a
+        #     hyphen (-), the user is only providing the name. Thus, we append `-*.whl`
+        #     to match the hyphen before the version, followed by anything else.
+        # - If the pattern is a glob, we ensure `*.whl` is appended if needed but
+        #   cowardly do not attempt glob parsing / project name normalization. The user
+        #   is on their own at that point.
         #
         # PEP 427: https://www.python.org/dev/peps/pep-0427/
-        pattern = pattern + ("*.whl" if "-" in pattern else "-*.whl")
-
-        return filesystem.find_files(wheel_dir, pattern)
+        # And: https://packaging.python.org/specifications/binary-distribution-format/
```

Comment:
For Pip / PyPA knowledgeable reviewers - what the heck is up with this by the way? I've always found it off-putting since the PEP banners have started appearing on some of the PyPA PEPs re-directing to these new standards pages. The fact that some PEPs get a new page and some don't and what process the ones with the new page go through to get updated is all mysterious feeling.